### PR TITLE
Code changes for zSystems (s390x)

### PR DESCRIPTION
**Pull request by shahidhs-ibm**
For s390x platform, we need following changes because:
1. The zSystems supports 'bogomips' instead of CPU clock speed (/proc/cpuinfo)
2. The retrieval of system uuid is done by file '/etc/machine-id'

With this patch, cAdvisor code can be built and executed on zSystems (s390x).

### CODE REVIEW

**Comment by vishh** [comment]
ok to test

File: utils/machine/machine.go [hunk 1]
```diff
@@ -241,3 +256,42 @@ func addNode(nodes *[]info.Node, id int) (int, error) {
...
+func getMachineArch() (string, error) {
+    	uname := syscall.Utsname{}
+    	err := syscall.Uname(&uname)
+    	if err != nil {
+        	return "", err
+    	}
+
+    	var arch string
+    	for _, val := range uname.Machine {
+        	arch += string(int(val))
```
Comment (vishh) [inline utils/machine/machine.go]: Does this have to be `Itoa()` ?

File: utils/sysfs/sysfs.go [hunk 2]
```diff
@@ -242,7 +243,11 @@ func (self *realSysFs) GetSystemUUID() (string, error) {
 			//If running on a KVM guest on Power then UUID is /proc/device-tree/vm,uuid
 			id, err = ioutil.ReadFile(path.Join(ppcDevTree, "vm,uuid"))
 			if err != nil {
-				return "", err
```
Comment (vishh) [inline utils/sysfs/sysfs.go]: This nested if loop isn't ideal. While you are here, can you refactor the code to not return `id` if any of the if conditions pass?

**Comment by vishh** [comment]
Just a couple of nits. Ping @rjnagal

File: utils/machine/machine.go (hunk 1 above)
Comment (shahidhs-ibm) [inline utils/machine/machine.go]: Here we are getting ASCII code from an array. We are converting this ASCII code into string and appending to form the machine hardware name (s390x for zSystem case). Hence we can't use 'Itoa' for this particular case.

File: utils/sysfs/sysfs.go (hunk 2 above)
Comment (shahidhs-ibm) [inline utils/sysfs/sysfs.go]: Updated the 'utils/sysfs/sysfs.go' with modified if-else statements. Please check and let me know if it looks ok.

**Comment by shahidhs-ibm** [comment]
Hi @vishh
Apologies for the delay in incorporating your review comments. Please check and let me know your comments.

**Comment by vishh** [comment]
LGTM. @shahidhs-ibm: All the tests are failing. Can you address the test issues?

**Comment by shahidhs-ibm** [comment]
@vishh 
It would be great if you can guide me on how to configure Travis for my zSystem specific build.

The cAdvisor code to build on s390x (zSystem) needs gccgo on s390x (Refer to link - https://github.com/linux-on-ibm-z/docs/wiki/Building-gccgo). This will need Travis to build gccgo first for s390x and then build cAdvisor.

**Comment by mvdan** [comment]
@shahidhs-ibm as a start, the gofmt checks are failing. Fix those first.

**Comment by shahidhs-ibm** [comment]
@mvdan Thanks for the info. I am new to Travis, so I will explore the configuration and will update once done.

**Comment by vishh** [comment]
ok to test

**Comment by shahidhs-ibm** [comment]
@vishh I'm working for IBM and covered as part of Google Corporate CLA (my id has been added as shahidhs.ibm@gmail.com). Please let me know if there is any additional info required.

**Comment by vishh** [comment]
ok to test

**Comment by shahidhs-ibm** [comment]
@vishh 
I tried to reproduce the failure of the 'default' build on my system but unable to reproduce it. Please let me know if the issue is related to my pull request or the environment on which build is running.

**Comment by vishh** [comment]
I have restarted the [test](http://104.197.39.211/job/cadvisor-e2e/2078/).

**Comment by shahidhs-ibm** [comment]
@vishh If the PR looks good to you, please let me know when it can be merged.

**Comment by vishh** [comment]
LGTM. Merging now. Thanks for the contribution @shahidhs-ibm!

**Comment by shahidhs-ibm** [comment]
Thanks @vishh ! :-)

**Comment by shahidhs-ibm** [comment]
@vishh 
I am not sure if this is right channel to discuss this topic but we would like to know if cAdvisor CI can be extended to support z system. Please let us know what support will be required from IBM side (we are aware on the H/w support).
Let me know if you want to start this discussion on some other channel.

**Comment by vishh** [comment]
You can setup a CI independently and post the build status to PRs on
github. Typically a bot account is setup on github for CI purposes. I can
grant the necessary privileges for the bot account to post build status.

**Comment by shahidhs-ibm** [comment]
Thanks vish for the information.
Can you please provide me more details on how the current CI infrastructure is like?

Also I need below details in order to setup the CI infrastructure on z Systems:
1. How frequent the CI run on average?
2. What is the recommended H/W and configuration requirement
3. Special S/W requirements (if any)

**Comment by vishh** [comment]
> 1. How frequent the CI run on average?

For every PR push. Take a look at ghprb on Jenkins.

> 2. What is the recommended H/W and configuration requirement

None AFAIK. Just build and test against your platform.

> 3. Special S/W requirements (if any)

None AFAIK.

**Comment by shahidhs-ibm** [comment]
Thanks Vishh.
We are in process of procuring the s390x VMs for the CI. We like to extend this CI discussion to include Kubernetes as well. We believe that the above information is also applicable for Kubernetes. Please let us know if this is not the case.
Also just curious on if there is any other way than ghprb to hook up the s390x VMs directly to cAdvisor and Kubernetes CI?

**Comment by vishh** [comment]
We cannot expand to test more platforms at the moment. If you are planning on testing Kubernetes, I believe there is an on-going effort to have distributed testing. I would recommend joining the kubernetes sig-testing.

**Comment by shahidhs-ibm** [comment]
Could you please let me know how to get in touch with 'sig-testing' group of Kubernetes? I checked their home page but not sure how to start the communication with them. Please guide me here.
Also please let me know about the next actions that you do in the CI once the build is success. My query is mainly related to the archiving source, binary etc post build.

**Comment by vishh** [comment]
https://kubernetes.slack.com/
<https://kubernetes.slack.com/messages/sig-testing/> ==>  #sig-testing
channel

**Pull request merged**